test/unit/dirty-cache-check.cmake

```diff
@@ -0,0 +1,58 @@
+include(${CPM_PATH}/CPM.cmake)
+include(${CPM_PATH}/testing.cmake)
+
+set(baseDir "${CMAKE_CURRENT_BINARY_DIR}/test_dirty_cache")
+set(childDir "${baseDir}/edgecase")
+
+find_package(Git REQUIRED)
+
+function(git_do dir)
+  execute_process(
+    COMMAND ${GIT_EXECUTABLE} ${ARGN}
+    RESULT_VARIABLE result
+    OUTPUT_VARIABLE status
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+    WORKING_DIRECTORY "${dir}"
+  )
+  if(result)
+    message(FATAL_ERROR "git ${ARGN} fail: ${result} ${status}")
+  endif()
+endfunction()
+
+file(MAKE_DIRECTORY "${baseDir}")
+
+file(WRITE "${baseDir}/draft.txt" "this is a test")
+message(STATUS "no git, a file")
+cpm_check_working_dir_is_clean(${baseDir} nogit_test)
+assert_truthy(nogit_test)
+
+git_do("${baseDir}" init -b main)
+message(STATUS "empty repo with file")
+cpm_check_working_dir_is_clean(${baseDir} emptygit_test)
+assert_falsy(emptygit_test)
+
+git_do("${baseDir}" add draft.txt)
+git_do("${baseDir}" commit -m "test change")
+message(STATUS "commit a change")
+cpm_check_working_dir_is_clean(${baseDir} onecommit_test)
+assert_truthy(onecommit_test)
```

Comment:
This is an interesting case I didn't consider before - do we want to accept adding new commits as "clean"?

I think the expected behavior might be to also validate that the `GIT_TAG` is as defined `CPMAddPackage`.

Reply:
makes sense. @TheLartians where should I look to extract the ```GIT_TAG``` of the ```CPMAddPackage``` call?

Reply:
Inside the CPMAddPackage call, the argument should be named `CPM_ARGS_GIT_TAG`, so I think you could just add an additional argument `GIT_TAG` to the `cpm_check_working_dir_is_clean` function and change the call to 

```cmake
       cpm_check_working_dir_is_clean(${download_directory} "${CPM_ARGS_GIT_TAG}" IS_CLEAN)
```